vllm/model_executor/layers/quantization/kernels/mixed_precision/xla.py

```diff
@@ -0,0 +1,85 @@
+from typing import Optional, Tuple
+
+import torch
+
+from vllm.scalar_type import scalar_types
+
+from .MPLinearKernel import MPLinearKernel, MPLinearLayerConfig
+
+XLA_SUPPORTED_QUANT_TYPES = [scalar_types.uint8b128]
+XLA_SUPPORTED_GROUP_SIZES = [-1]
+
+
+class XLAMixedPrecisionLinearKernel(MPLinearKernel):
+    """
+    XLAMixedPrecisionLinearKernel: WNA16 for TPU
+    
+    Kernel definition:
+        - https://github.com/pytorch/xla/blob/v2.5.0-rc9/torch_xla/experimental/xla_quantized_matmul.py#L78
+
+    Supported:
+        - w8a16 symmetric channelwise
+
+    Currently unsupported:
+        - w8a16
+        - w8a16 grouped
+        - w4a16 
+        - asymmetric
+        - activation_reordering
+    """
+
+    @classmethod
+    def get_min_capability(cls) -> Optional[int]:
+        return None
+
+    @classmethod
+    def can_implement(cls,
+                      c: MPLinearLayerConfig) -> Tuple[bool, Optional[str]]:
+
+        if c.zero_points:
+            return False, "Zero points currently not supported by XLA"
+
+        if c.weight_type not in XLA_SUPPORTED_QUANT_TYPES:
+            return False, f"Quant type ({c.weight_type}) not supported by XLA"\
+                          f" , supported types are: {XLA_SUPPORTED_QUANT_TYPES}"
+
+        if c.group_size not in XLA_SUPPORTED_GROUP_SIZES:
+            return False, f"Group size ({c.group_size}) not supported by "\
+                            "XLA, supported group sizes are: "\
+                            f"{XLA_SUPPORTED_GROUP_SIZES}"
+
+        if c.has_g_idx:
+            return False, "Activation reordering is not supported by XLA"
+
+        return True, None
+
+    # note assumes that
+    #  `weight_packed` is: {input_dim = 0, output_dim = 1, packed_dim = 0}
+    #  `weight_scale` is: {input_dim = 0, output_dim = 1}
+    def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
```

Comment:
qq: does the post-processing happen on host or device? If it's on device we need to call `xm.mark_step` at the end, similar to https://github.com/vllm-project/vllm/blob/ec10cb8511b7e30b8ff86caab2e4272ff3ceddca/vllm/model_executor/model_loader/loader.py#L357.

If it's not called, XLA will record the post-processing computation along with the model execution.

Reply:
Yes, I think this needs to be used. The repacking logic is not quite working yet and is causing oom

Reply:
@robertgshaw2-neuralmagic do you know if compressed tensor stores as `uint8b128` or `int8`? It appears from the existing code `uint8b128`, if that is the case to go from packed `uint8b128` we'd have to do something like:
```
        def transform_w_int8(x):
             # reinterpet_cast the packed int32 weights as int8
             # convert to [int, out] -> [out, int]
             return (x.view(dtype=torch.uint8).to(torch.int16) - 128).to(torch.int8).t()
```

basically if I had the values:
`[  0,   1,  -1,  -2,   2,  17, -28,  72]`
for `uint8b128` these would be encoded as:
```
tensor([128, 129, 127, 126, 130, 145, 100, 200], dtype=torch.uint8)
```
for `int8` 
```
tensor([  0,   1,  -1,  -2,   2,  17, -28,  72], dtype=torch.int8)
```

(so to convert you do `(x.view(dtype=torch.uint8).to(torch.int16) - 128).to(torch.int8)`, i.e. reinterpret to `torch.uint8` then upcast to `int16` so we can subtract the bias, subtract the 128 bias then cast to `int8`)

e.g.
```
>>> x
tensor([2122285440, -932933246], dtype=torch.int32)
>>> x.view(torch.uint8)
tensor([128, 129, 127, 126, 130, 145, 100, 200], dtype=torch.uint8)
>>> (x.view(dtype=torch.uint8).to(torch.int16) - 128).to(torch.int8)
tensor([  0,   1,  -1,  -2,   2,  17, -28,  72], dtype=torch.int8)
```


(not sure why it would go OOM though)

(cc @dsikka @mgoin , do you know if compressed tensor is stored unsigned with a bias?)